### ISSUE TITLE
Shouldn't no-invalid-this rule flag use of this inside nested functions in method?

### ISSUE THREAD

**Issue by rtm** [palantir/tslint#3625]
### Bug Report

- __TSLint version__: 5.8.0
- __TypeScript version__: 2.6.1
- __Running TSLint via CLI

#### TypeScript code being linted

```ts
class Foo {
  method() { 
    function nested() { this.foo = 42; }
  }
}
```

with `tslint.json` configuration:

```json
"no-invalid-this": true
```

#### Actual behavior

No error is reported.

#### Expected behavior

I would hope that the use of `this` would be reported.

**Comment by JoshuaKGoldberg** [comment]
Yup, this is a bug. The rule's walker function keeps track of an `inClass` boolean that's set to `true` when it enters a class and `false` when it leaves. It should also reset the variable when it enters or leaves a function declaration.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
